### PR TITLE
Fix SmacPlanner memory leaks and costmap downsampler test

### PR DESCRIPTION
**Pull request by naiveHobo**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | - |
| Primary OS tested on | Ubuntu 18.04 |
| Robotic platform tested on | TurtleBot3 Gazebo Simulation |

---

## Description of contribution in a few bullet points

* Fix `shared_ptr` cyclic dependencies in `SmacPlanner` and `CostmapDownsampler` as per #1900
* Refactor `CostmapDownsampler` to have `on_` type functions to configure, activate, deactivate, and cleanup costmap publisher stuff
* Fix `CostmapDownsampler` test which has been failing on Cyclone DDS (#2043)

### CODE REVIEW

File: smac_planner/src/costmap_downsampler.cpp
Comment (SteveMacenski) [inline smac_planner/src/costmap_downsampler.cpp:93]: This should still only publish when someone is listening. Maybe this is something that the costmap publisher object can do? It itself has a publisher under the hood and the publisher can check for the number of subscribers rather than the node

File: smac_planner/test/test_costmap_downsampler.cpp
Comment (SteveMacenski) [inline smac_planner/test/test_costmap_downsampler.cpp:33]: Can you explain your thinking a little on this test node? I'm not sure I see the benefit of these changes

**Comment by SteveMacenski** [comment]
Overall, looks good to me, just a couple things to clarify

File: smac_planner/src/costmap_downsampler.cpp
Comment (naiveHobo) [reply smac_planner/src/costmap_downsampler.cpp:93]: Exactly, `Costmap2DPublisher::publishCostmap()` already checks for subscription count and publishes only when someone is listening.

Comment (SteveMacenski) [reply smac_planner/src/costmap_downsampler.cpp:93]: Oh nice, ok

File: smac_planner/test/test_costmap_downsampler.cpp
Comment (SteveMacenski) [reply smac_planner/test/test_costmap_downsampler.cpp:33]: This is the last point of just trying to understand the need behind the changes :smile:

Comment (naiveHobo) [reply smac_planner/test/test_costmap_downsampler.cpp]: Ah yes on second look seems like those changes were redundant. I was under the impression that the test was failing on cyclonedds because the parent `nav2_util::LifecycleNode` needed to be in the configuring lifecycle transition state when the downsampler is configured. I've reverted the changes in the latest commit and everything seems to work now.

**Review by SteveMacenski — APPROVED** [review]
Waiting for CI to stand up again from the gazebo ros pkgs issue before merging. Marking as "merge" as it will be once up again

**Comment by SteveMacenski** [comment]
Please rebase / merge in main, CI has been fixed

**Comment by naiveHobo** [comment]
@SteveMacenski is this okay to merge? This should close #2043 I believe, all the recurring CI test failures have been fixed.

**Comment by SteveMacenski** [comment]
I don't think it would close #2043 until we see cyclone jobs succeeding in nightly for a few days.

**Pull request merged**
